src/trace/implementations/merge_batcher_col.rs

```diff
@@ -0,0 +1,353 @@
+                    }
+                    keep.copy(datum);
+                }
+                else {
+                    builder.push((key.clone(), val.clone(), time.clone(), diff.clone()));
+                }
+                if let Some(last) = last {
+                    assert!(last <= datum);
+                }
+                last = Some(datum);
+            }
+            // Recycling buffer.
+            buffer.clear();
+            self.sorter.recycle(buffer);
+        }
+
+        // Finish the kept data.
+        if !keep.is_empty() {
+            kept.push(keep);
+        }
+        if !kept.is_empty() {
+            self.sorter.push_list(kept);
+        }
+
+        // Drain buffers (fast reclamation).
+        self.sorter.clear_stash();
+
+        let seal = builder.done(self.lower.clone(), upper.clone(), Antichain::from_elem(<B::Time as timely::progress::Timestamp>::minimum()));
+        self.lower = upper;
+        seal
+    }
+
+    // the frontier of elements remaining after the most recent call to `self.seal`.
+    fn frontier(&mut self) -> timely::progress::frontier::AntichainRef<B::Time> {
+        self.frontier.borrow()
+    }
+}
+
+struct TimelyStackQueue<T: Columnation> {
+    list: TimelyStack<T>,
+    head: usize,
+}
+
+impl<T: Columnation + 'static> TimelyStackQueue<T> {
+
+    fn new() -> Self { TimelyStackQueue::from(Default::default()) }
+
+    fn pop(&mut self) -> &T {
+        self.head += 1;
+        &self.list[self.head - 1]
+    }
+
+    fn peek(&self) -> &T {
+        &self.list[self.head]
+    }
+
+    fn from(list: TimelyStack<T>) -> Self {
+        TimelyStackQueue {
+            list,
+            head: 0,
+        }
+    }
+
+    fn done(mut self) -> TimelyStack<T> {
+        self.list.clear();
+        self.list
+    }
+
+    fn len(&self) -> usize { self.list.len() - self.head }
+
+    fn is_empty(&self) -> bool { self.head == self.list.len() }
+
+    /// Return an iterator over the remaining elements.
+    fn iter(&self) -> impl Iterator<Item=&T> + Clone + ExactSizeIterator {
+        self.list.iter().skip(self.head)
+    }
+}
+
+struct MergeSorterColumnation<D: Ord+Columnation, T: Ord+Columnation, R: Semigroup+Columnation> {
+    queue: Vec<Vec<TimelyStack<(D, T, R)>>>,    // each power-of-two length list of allocations.
+    stash: Vec<TimelyStack<(D, T, R)>>,
+    pending: Vec<(D, T, R)>,
+}
+
+impl<D: Ord+Clone+Columnation+'static, T: Ord+Clone+Columnation+'static, R: Semigroup+Columnation+'static> MergeSorterColumnation<D, T, R> {
+
+    const BUFFER_SIZE_BYTES: usize = 64 << 10;
+
+    /// Buffer size (number of elements) to use for new/empty buffers.
+    const fn buffer_size() -> usize {
+        let size = std::mem::size_of::<(D, T, R)>();
+        if size == 0 {
+            Self::BUFFER_SIZE_BYTES
+        } else if size <= Self::BUFFER_SIZE_BYTES {
+            Self::BUFFER_SIZE_BYTES / size
+        } else {
+            1
+        }
+    }
```

Comment:
Also, and worth discussing: I think this size has roughly nothing to do with what `lg_alloc` will have to absorb? It indicates how large a `Vec<(D, T, R)>` will be, but the container will instead hold on to the spilled data, which can be arbitrarily larger (e.g. if `D` is `mz_repr::Row`), or arbitrarily smaller. It might be a good set-up for when you are able to back the `Vec` itself in `lg_alloc`, but .. is it really doing the right thing until then?

Seems not terrible ofc, but idk if we are sure that the backing allocations will be able to be paged out using this logic. Perhaps it's fine, in that it ensures that we take a large enough bite that the `Vec` would be that large, and if the alloc is smaller then we have at most 2x the `Vec` in memory (it plus a bit less in the region, in the worst case), which is decent insurance to provide (vs a smaller `Vec` which might have proportionally more in the region, and still not spill out).

Reply:
I added a `pending_buffer_size` function that returns `2*Self::buffer_size()` so that pending contains enough data to fill a whole buffer. I need to test the performance impact of this, but it seems like the right change to do.